docs/source/spec.rst

```diff
@@ -20,7 +20,12 @@ A **Dask graph** is a dictionary mapping **keys** to **computations**:
     'w': (sum, ['x', 'y', 'z']),
     'v': [(sum, ['w', 'z']), 2]}
 
-A **key** is any hashable value that is not a **task**:
+
+A **key** is either:
+
+- a string, or
+- a tuple of one or more elements, whose first element is a string and the remainder are
+  arbitrary hashables:
```

Comment:
Changing the definition of a key here seems unnecessary, and the old definition remains true.  In practice, keys are typically strings or tuples as you describe, but this isn't enforced from what I see.  For example, perhaps somebody paranoid would prefer to build a dask graph using objects of a `Key` class that they create.  I haven't seen this done, but the previous spec definition allows for this.  Dask collections (including custom collections) are not the only way to create dask graphs.

Updating the specification for keys of collections (the bulk of this PR) is a good idea.